bundler/lib/bundler/dsl.rb

```diff
@@ -45,6 +46,15 @@ def eval_gemfile(gemfile, contents = nil)
       @gemfiles << expanded_gemfile_path
       contents ||= Bundler.read_file(@gemfile.to_s)
       instance_eval(contents.dup.tap{|x| x.untaint if RUBY_VERSION < "2.7" }, gemfile.to_s, 1)
+
+      unless @global_source_defined
```

Comment:
`instance_eval` will load the Gemfile source so in the next step we will raise a warning if no global source has been defined. 

Let me know what you think about the warning message.